staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -432,3 +443,42 @@ func (r *Reflector) setLastSyncResourceVersion(v string) {
 	defer r.lastSyncResourceVersionMutex.Unlock()
 	r.lastSyncResourceVersion = v
 }
+
+// relistResourceVersion determines the resource version the reflector should list or relist from.
+// Returns either the lastSyncResourceVersion so that this reflector will relist with a resource
+// versions no older than has already been observed in relist results or watch events, or, if the last relist resulted
+// in an HTTP 410 (Gone) status code, returns "" so that the relist will use the latest resource version available in
+// etcd via a quorum read.
+func (r *Reflector) relistResourceVersion() string {
+	r.lastSyncResourceVersionMutex.RLock()
+	defer r.lastSyncResourceVersionMutex.RUnlock()
+
+	if r.isLastSyncResourceVersionGone {
```

Comment:
Only if WatchListPageSize > 0

Reply:
Not sure I follow. Is the goal to avoid any unpaginated RV="" list request?

If we've already received a 410 Gone, then we can't do a RV="n" request. It would just return Gone again. If we do RV="0", we can get a stale read, which is what we're trying to avoid.

Reply:
No, I mean you can create a Reflector that doesn't use paging.  So any logic related to paging has to be constrained to `r.WatchListPageSize > 0`

Reply:
I think I understand.

I've rewritten the comment explaining the fallback from paginated to full list:

https://github.com/kubernetes/kubernetes/pull/83520/files#diff-df2b34adba86da0485290415620d737aR218

Does that make sense? I believe the logic is correct.

Reply:
Yeah, I think so.